lib/Date/Utility.pm

```diff
@@ -781,32 +778,85 @@ Returns a TimeInterval which represents the difference between UTC and the time
 
 =cut
 
-sub timezone_offset {
-    my ($self, $timezone) = @_;
-
-    my $dt = DateTime->from_epoch(
-        epoch     => $self->{epoch},
-        time_zone => $timezone
-    );
-
-    return Time::Duration::Concise::Localize->new(interval => $dt->offset);
-}
-
 =head2 is_dst_in_zone
 
 Returns a boolean which indicates whether a certain zone is in DST at the given epoch
 
 =cut
 
-sub is_dst_in_zone {
-    my ($self, $timezone) = @_;
+{
+    use DateTime;
+    use DateTime::TimeZone;
 
-    my $dt = DateTime->from_epoch(
-        epoch     => $self->{epoch},
-        time_zone => $timezone
-    );
+    my $bignum = 20000000;
+
+    my %cache;
+    my $cache_for = sub {
+        my $tm = shift;
+        my $tzname = shift;
+        my $k = int $tm/$bignum;
+
+        if (my $val = $cache{"$k $tzname"}) {
+            return $val;
+        }
+
+        my $z = DateTime::TimeZone->new(name => $tzname);
+        my $start_of_interval = $k * $bignum;
+        my $dt = DateTime->from_epoch(epoch => $start_of_interval);
+        my $rdoff = $dt->utc_rd_as_seconds - $start_of_interval;
+
+        my ($span_start, $span_end, undef, undef, $off, $is_dst, $name) = @{$z->_span_for_datetime(utc => $dt)};
+        $_ -= $rdoff for ($span_start, $span_end);
+
+        my @val = ([$span_start, $span_end, $off, $is_dst, $name]);
+
+        while ($span_end < ($k+1) * $bignum) {
+            $dt = DateTime->from_epoch(epoch => $span_end);
+
+            ($span_start, $span_end, undef, undef, $off, $is_dst, $name) = @{$z->_span_for_datetime(utc => $dt)};
+            $_ -= $rdoff for ($span_start, $span_end);
+
+            push @val, [$span_start, $span_end, $off, $is_dst, $name];
```

Comment:
Any reason not to use a hash here? It would make the code a bit easier to follow.